extensions/ql-vscode/src/telemetry.ts

```diff
@@ -0,0 +1,159 @@
+
+export enum CommandCompletion {
+  Success = 'Success',
+  Failed = 'Failed',
+  Cancelled = 'Cancelled'
+}
+
+export async function initializeTelemetry(extension: Extension<any>, ctx: ExtensionContext): Promise<void> {
+  await requestTelemetryPermission(ctx);
+
+  registerListener(extension, ctx);
+  if (reporter) {
+    reporter.dispose();
+    reporter = undefined;
+  }
+  if (ENABLE_TELEMETRY.getValue<boolean>()) {
+    reporter = new TelemetryReporter(
+      extension.id,
+      extension.packageJSON.version,
+      key,
+      /* anonymize stack traces */ true
+    );
+
+    // add a telemetry processor to delete unwanted properties
+    const client = (reporter as any).appInsightsClient as appInsights.TelemetryClient;
+    if (client) {
+      client.addTelemetryProcessor((envelope) => {
+        delete envelope.tags['ai.cloud.roleInstance'];
+        delete (envelope.data as any)?.baseData?.properties?.['common.remotename'];
+
+        return true;
+      });
+
+      // add a telemetry processor to log if requested
+      client.addTelemetryProcessor((envelope) => {
+        if (LOG_TELEMETRY.getValue<boolean>()) {
+          logger.log(`Telemetry: ${JSON.stringify(envelope)}`);
+        }
+        return true;
+      });
+    }
+
+    ctx.subscriptions.push(reporter);
+  }
+}
+
+export function sendCommandUsage(name: string, executionTime: number, error?: Error) {
+  if (!reporter) {
+    return;
+  }
+  const status = !error
+    ? CommandCompletion.Success
+    : error instanceof UserCancellationException
+      ? CommandCompletion.Cancelled
+      : CommandCompletion.Failed;
+
+  const isCanary = (!!CANARY_FEATURES.getValue<boolean>()).toString();
+
+  reporter.sendTelemetryEvent(
+    'command-usage',
+    {
+      name,
+      status,
+      isCanary
+    },
+    { executionTime }
+  );
+
+  // if this is a true error, also report it
+  if (status === CommandCompletion.Failed && error) {
+    const redactedError = {
+      name: error.name,
+      stack: error.stack,
+      message: '<MESSAGE REDACTED>'
+    };
+    reporter.sendTelemetryException(
+      redactedError,
+      {
+        type: 'command-usage',
+        name,
+        status,
+        isCanary
+      },
+      { executionTime }
+    );
+  }
+}
+
+function registerListener(extension: Extension<any>, ctx: ExtensionContext) {
+  if (!listener) {
+    listener = workspace.onDidChangeConfiguration(async e => {
+      if (e.affectsConfiguration('codeQL.telemetry.enableTelemetry')) {
+        await initializeTelemetry(extension, ctx);
+      }
+
+      // TODO: re-request telemetry so that users can see the dialog again.
+      // Re-request if codeQL.canary is being set to `true` and telemetry
+      // is not currently enabled.
+      if (
+        e.affectsConfiguration('codeQL.canary') &&
+        CANARY_FEATURES.getValue() &&
+        !ENABLE_TELEMETRY.getValue()
+      ) {
+        await ctx.globalState.update('telemetry-request-viewed', false);
```

Comment:
Either pull the string key out into a constant, or create a helper method that takes `ctx` and a boolean.

Reply:
sure